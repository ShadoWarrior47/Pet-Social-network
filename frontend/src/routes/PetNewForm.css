/* General container styling */
.PetNewForm-container div {
    max-width: 600px !important;
    margin: 0px auto;
    padding: 10px;
    font-family: Arial, sans-serif;
    min-width: 500px;
}

.PetNewForm-container form {
    display: flex;
    flex-direction: column;
    gap: 10px;
    padding: 10px;
    background-color: #f9f9f9;
    border: 1px solid #ddd;
    border-radius: 5px;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1);
}

.PetNewForm-container h1 {
    text-align: center;
    margin-bottom: 20px;
    font-size: 24px;
    color: #333;
}

.PetNewForm-container input {
    width: 100%;
    padding: 5px;
    font-size: 16px;
    border: 1px solid #ccc;
    border-radius: 4px;
    box-shadow: inset 0 1px 3px rgba(0, 0, 0, 0.1);
    outline: none;
    transition: border-color 0.3s;
}

.PetNewForm-container input:focus {
    border-color: #007bff;
}

.PetNewForm-container label {
    font-size: 14px;
    font-weight: bold;
    margin-bottom: 2px;
    color: #555;
    text-align: center;
}

.PetNewForm-container button {
    padding: 10px 20px;
    font-size: 16px;
    color: white;
    background-color: #007bff;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s, transform 0.2s;
    width: 150px;
    margin: 0px auto;
}

.PetNewForm-container button:hover {
    background-color: #0056b3;
}

.PetNewForm-container button:active {
    transform: scale(0.98);
}