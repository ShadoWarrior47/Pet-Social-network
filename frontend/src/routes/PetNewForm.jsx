import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import PetApi from "../PetApi";
import "./PetNewForm.css";

function PetNewForm({ createPet, currentUser }) {
  const navigate = useNavigate();

  const INITIAL_STATE = {
    name: "",
    type: "",
    breed: "",
    age: "",
    bio: "",
    photoUrl: "",
  };

  const [formData, setFormData] = useState(INITIAL_STATE);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const petData = { ...formData, ownerId: currentUser.id, age: parseInt(formData.age, 10) || 0 };
      await createPet(petData); // `createPet` makes the API call
      navigate("/pets"); // Navigate to the pet listing or relevant page
    } catch (err) {
      console.error("Error creating pet:", err);
      alert("Failed to create pet. Please try again.");
    }
  };

  return (
    <div className="PetNewForm-container">
      <h1>Add a New Pet</h1>
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="name">Name:</label>
          <input
            id="name"
            name="name"
            type="text"
            placeholder="Pet's Name"
            value={formData.name}
            onChange={handleChange}
            required
          />
        </div>
        <div>
          <label htmlFor="type">Type:</label>
          <input
            id="type"
            name="type"
            type="text"
            placeholder="Pet's Type (e.g., Dog, Cat)"
            value={formData.type}
            onChange={handleChange}
            required
          />
        </div>
        <div>
          <label htmlFor="breed">Breed:</label>
          <input
            id="breed"
            name="breed"
            type="text"
            placeholder="Pet's Breed"
            value={formData.breed}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="age">Age:</label>
          <input
            id="age"
            name="age"
            type="number"
            placeholder="Pet's Age"
            value={formData.age}
            onChange={handleChange}
            min="0"
          />
        </div>
        <div>
          <label htmlFor="bio">Bio:</label>
          <textarea
            id="bio"
            name="bio"
            placeholder="Tell us about your pet"
            value={formData.bio}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="photoUrl">Photo URL:</label>
          <input
            id="photoUrl"
            name="photoUrl"
            type="text"
            placeholder="Photo URL"
            value={formData.photoUrl}
            onChange={handleChange}
          />
        </div>
        <button type="submit">Add Pet</button>
      </form>
    </div>
  );
}

export default PetNewForm;
