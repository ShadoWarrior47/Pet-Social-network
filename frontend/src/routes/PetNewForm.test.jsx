import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import PetNewForm from "./PetNewForm";
import { vi } from "vitest";

it("renders without crashing", function () {
    render(
        <MemoryRouter>
            <PetNewForm createPet={() => { }} currentUser={{ id: 1, username: "testuser" }} />
        </MemoryRouter>
    );
});

it("matches snapshot", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <PetNewForm createPet={() => { }} currentUser={{ id: 1, username: "testuser" }} />
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

it("calls createPet function on form submission", async function () {
    const createPetMock = vi.fn();
    const { getByLabelText, getByText } = render(
        <MemoryRouter>
            <PetNewForm createPet={createPetMock} currentUser={{ id: 1, username: "testuser" }} />
        </MemoryRouter>
    );

    // Select form inputs
    const nameInput = getByLabelText("Name:");
    const typeInput = getByLabelText("Type:");
    const breedInput = getByLabelText("Breed:");
    const ageInput = getByLabelText("Age:");
    const bioInput = getByLabelText("Bio:");
    const photoUrlInput = getByLabelText("Photo URL:");
    const submitButton = getByText("Add Pet");

    // Simulate user input
    fireEvent.change(nameInput, { target: { value: "Buddy" } });
    fireEvent.change(typeInput, { target: { value: "Dog" } });
    fireEvent.change(breedInput, { target: { value: "Golden Retriever" } });
    fireEvent.change(ageInput, { target: { value: "3" } });
    fireEvent.change(bioInput, { target: { value: "A friendly dog" } });
    fireEvent.change(photoUrlInput, { target: { value: "http://example.com/photo.jpg" } });

    // Submit the form
    fireEvent.click(submitButton);

    // Assert that createPet was called with correct data
    expect(createPetMock).toHaveBeenCalledWith({
        name: "Buddy",
        type: "Dog",
        breed: "Golden Retriever",
        age: 3,
        bio: "A friendly dog",
        photoUrl: "http://example.com/photo.jpg",
        ownerId: 1,
    });
});
