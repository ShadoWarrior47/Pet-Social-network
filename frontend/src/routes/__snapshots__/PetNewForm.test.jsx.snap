// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PetNewForm-container"
  >
    <h1>
      Add a New Pet
    </h1>
    <form>
      <div>
        <label
          for="name"
        >
          Name:
        </label>
        <input
          id="name"
          name="name"
          placeholder="Pet's Name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="type"
        >
          Type:
        </label>
        <input
          id="type"
          name="type"
          placeholder="Pet's Type (e.g., Dog, Cat)"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="breed"
        >
          Breed:
        </label>
        <input
          id="breed"
          name="breed"
          placeholder="Pet's Breed"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="age"
        >
          Age:
        </label>
        <input
          id="age"
          min="0"
          name="age"
          placeholder="Pet's Age"
          type="number"
          value=""
        />
      </div>
      <div>
        <label
          for="bio"
        >
          Bio:
        </label>
        <textarea
          id="bio"
          name="bio"
          placeholder="Tell us about your pet"
        />
      </div>
      <div>
        <label
          for="photoUrl"
        >
          Photo URL:
        </label>
        <input
          id="photoUrl"
          name="photoUrl"
          placeholder="Photo URL"
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Add Pet
      </button>
    </form>
  </div>
</DocumentFragment>
`;
